spring-cloud-skipper-server-core/src/test/java/org/springframework/cloud/skipper/server/statemachine/StateMachineTests.java

```diff
@@ -419,12 +420,24 @@ public void testUpgradeCancelWhileCheckingApps() throws Exception {
 					.step()
 						.sendEvent(message2)
 						.expectStateChanged(2)
-						.expectStateEntered(SkipperStates.UPGRADE_CANCEL,
-								SkipperStates.INITIAL)
+						// for now need to do a trick to wait this later
```

Comment:
@jvalkeal Could you elaborate on this specific change? I am not sure I understand the reasoning (especially with Boot2 changes).

Reply:
There is a this timer loop in a machine which allows to break out from it if we'd do a cancel(which btw is currently not really even used but machine model allows it). There were some changes in a statemachine in 2.x line working little differently what comes for these timers, which then cause these loops to work differently. As I wanted to solely rely on spring-statemachine-test package to execute tests, I now noticed that test step executions are missing some functionality to conditionally wait something to happen. That's why I needed to modify this test to wait it to eventually end up back to INITIAL state. This test case was originally a bit of hack as you get trouble with threads and timers.

But as I said, what this test is testing is not currently used in a skipper as we don't support cancelling an upgrade while used machine theoretically supports it.